cpp/src/arrow/record_batch_test.cc

```diff
@@ -705,17 +705,12 @@ TEST_F(TestRecordBatch, ToTensorSupportedNaN) {
   std::vector<int64_t> shape = {9, 2};
   const int64_t f32_size = sizeof(float);
   std::vector<int64_t> f_strides = {f32_size, f32_size * shape[0]};
-  std::vector<float> f_values = {
-      static_cast<float>(NAN), 2,  3,  4,  5, 6, 7, 8, 9, 10, 20, 30, 40,
-      static_cast<float>(NAN), 60, 70, 80, 90};
-  auto data = Buffer::Wrap(f_values);
-
-  std::shared_ptr<Tensor> tensor_expected;
-  ASSERT_OK_AND_ASSIGN(tensor_expected, Tensor::Make(float32(), data, shape, f_strides));
+  std::shared_ptr<Tensor> tensor_expected = TensorFromJSON(
+      float32(), shape,
+      "[NaN, 2,  3,  4,  5, 6, 7, 8, 9, 10, 20, 30, 40, NaN, 60, 70, 80, 90]", f_strides);
```

Comment:
I think for most cases it'd be preferable to infer both shape and strides from the JSON, which can directly represent multidimensional arrays:
```suggestion
  auto tensor_expected = TensorFromJSON(
      float32(), R"( [
        [NaN, 2],
        [  3,  4],
        [  5,  6],
        [  7,  8],
        [  9, 10],
        [ 20, 30],
        [ 40, NaN],
        [ 60, 70],
        [ 80, 90]
      ] )");
```

Reply:
Aha, interesting. And a bit more complicated for only testing purposes?
@rok what do you think?

Reply:
Passing shapes implicitly is ok, but we would still need to explicitly pass strides (or dimension permutations) for cases where we have non-row-major strides. Maybe instantiated Tensor's strides can be altered, but that seems error prone and not really helpful. @bkietz

Reply:
I'd say allowing for optional dimension permutations is the best way to support non-row-major strides (my first thought for handling that was an enum but I like the permutation idea better).

Reply:
For testing purposes, I think it might be better to very explicitly create the values as they are stored in the tensor, instead of implicitly reshuffling them (which is essentially what would happen if you present the data as a Nd json array, but then combined that with non-trivial permutation or strides). IMO that would make the tests only harder to interpret, so I would go for the simpler option (i.e. current PR)

Reply:
I slightly lean towards passing shape explicitly (current proposal), but am ok with either approach as long as there is a way to provide strides/permutations explicitly.

Reply:
> it might be better to very explicitly create the values as they are stored in the tensor

ArrayFromJSON has no way to specify many valid arrow arrays, for example arrays with non-zero/empty values under null bits or the much larger ambiguity of offsets in Utf8. Constructing these directly is of course still possible when necessary using other functions, as in any test written before the addition of a FromJSON factory. However in the majority of test cases such flexibility would be unnecessary and distracting as we only need an obvious, self-contained, and maximally terse literal value. IMO writing a nested Json array in order to represent a nested Arrow array is clearly preferable and doing otherwise misses an opportunity for clarity in the most common cases; explicit shape/strides are still easy enough to specify using existing functions including ArrayFromJSON.

Reply:
I would agree with you @bkietz on the topic of the nested Arrow arrays. But in the case of Arrow Tensors the difference between logical and physical layout already makes things hard to visualise and Nd json would make it even more complicated. I propose to leave it as is with explicitly creating values as they are stored. If we later see the benefit of having Nd json where shape is inferred, I am happy to do a follow-up adding a second, "RowMajor TensorFromJSON" for example (to make things a bit clearer).

Thank you @rok for the info about `ComputeStrides` helper! Will use it and add permutation as a possible input also.